agent/dns.go

```diff
@@ -264,6 +264,28 @@ func (d *DNSServer) ListenAndServe(network, addr string, notif func()) error {
 	return d.Server.ListenAndServe()
 }
 
+func (d *DNSServer) Shutdown() {
+	if d.Server != nil {
+		d.logger.Info("Stopping server",
+			"protocol", "DNS",
+			"address", d.Server.Addr,
+			"network", d.Server.Net,
+		)
+		err := d.Server.Shutdown()
+		if err != nil {
+			d.logger.Error("Error stopping DNS server", "error", err)
+		}
+	}
+}
+
+// GetAddr is a function to return the server address if is not nil.
+func (d *DNSServer) GetAddr() string {
+	if d.Server != nil {
+		return d.Server.Addr
+	}
+	return ""
+}
+
```

Comment:
These just move functionality from `agent.go` into the existing `DNSServer` implementation.